### PR TITLE
Add MidiMe model

### PR DESCRIPTION
**Pull request by notwaldorf**

### CODE REVIEW

File: music/src/core/data_test.ts [hunk 1]
```diff
@@ -261,8 +259,8 @@ test('Test TrioConverter', (t: test.Test) => {
   const trioTensor = trioConverter.toTensor(TRIO_NS);
   t.deepEqual(trioTensor.shape, [32, 90 + 90 + 512]);
   const value = tf.tidy(
-      () => trioTensor.sum(1).equal(tf.scalar(3, 'int32')).sum().arraySync() as
-          number);
+      () =>
+          trioTensor.sum(1).equal(tf.scalar(3, 'int32')).sum().arraySync());
```
Comment (notwaldorf) [inline music/src/core/data_test.ts]: these changes shouldn't be here, they're from a bad merge.

File: music/demos/midi_me.html [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+<body>
+  <h1>MidiME</h1>
+  <p><code>MidiME</code> is a variational autoencoder that lets you personalize
+    your own <code>MusicVAE</code> model with just a little data, so that samples
+    from <code>MidiME</code> sound closer to your provided data.</p>
+  <h2>Input data</h2>
+  <section>
+    <p>This is what you want the samples to sound like. Note that MusicVAE is a
+      monophonic model, so if you use a polyphonic MIDI file, you're not
+      guaranteed to reconstruct the "main" melody.</p>
```
Comment (jesseengel) [inline music/demos/midi_me.html]: I tried out the demo and it's fun! You may want to either let someone upload several midi files, or add a little text to let people know they should have a longer midi file (at least something more than a couple of bars) so that it doesn't just play that one thing over and over.

Alternatively, having an example midi for people to try without finding their own could do the trick.

File: music/demos/common_graph.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
+/**
+ * @license
+ * Copyright 2018 Google Inc. All Rights Reserved.
```
Comment (adarob) [inline music/demos/common_graph.ts]: https://bit.ly/2FDiFg6

File: music/demos/midi_me.html (hunk 2 above)
Comment (adarob) [reply music/demos/midi_me.html]: +1

File: music/demos/midi_me.html [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    <label class="button" id="fileBtn" disabled>
+      Load midi file
+      <input type="file" id="fileInput">
+    </label>
+    <p><code id="input"></code>
+  </section>
+
+  <h2>Training</h2>
+  <p>The MidiMe model works by training a variational autoencoder in the
+    browser. The more you train for, the better the input reconstruction will be.
```
Comment (adarob) [inline music/demos/midi_me.html]: The more *steps* you train for

File: music/demos/midi_me.html (hunk 2 above)
Comment (adarob) [reply music/demos/midi_me.html]: You can also add something like "Try training on a single full song to get outputs that sound like variations on it, or train on multiple songs to get samples that combine various characteristics of them."

File: music/src/music_vae/midi_me.ts [hunk 5]
```diff
@@ -0,0 +1,299 @@
+import * as tf from '@tensorflow/tfjs';
```
Comment (adarob) [inline music/src/music_vae/midi_me.ts]: add copyright and file description

File: music/demos/midi_me.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
+import * as tf from '@tensorflow/tfjs';
```
Comment (adarob) [inline music/demos/midi_me.ts]: add copyright and file description

File: music/demos/midi_me.html [hunk 7]
```diff
@@ -0,0 +1,77 @@
+
+<!-- Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (adarob) [inline music/demos/midi_me.html]: https://bit.ly/2FDiFg6

File: music/src/music_vae/midi_me.ts [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+import * as logging from '../core/logging';
+export {MidiMe};
+
+/**
+ * Class for transforming a single variable Gaussian distribution into a
```
Comment (adarob) [inline music/src/music_vae/midi_me.ts]: Huh? This description looks wrong.

File: music/src/music_vae/midi_me.ts [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+  }
+}
+
+/**
+ * An interface for providing configurable properties to the MidiMe model.
+ * @param input_size The shape of the VAE input. Since the inputs to this
+ * VAE are actually latent vectors from MusicVAE, then this number should be
+ * equal to the number of latent variables used by MusicVAE (`zDims`). The
+ * default is 256.
+ * @param output_size The size of the output. The default is 4.
```
Comment (adarob) [inline music/src/music_vae/midi_me.ts]: By output do you mean the number of dimensions of the MidiMe latent vector? Please use a clearer name and description.

File: music/src/music_vae/midi_me.ts [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+ * @param encoder_layers The shape of the layers in the Encoder network.
+ * @param decoder_layers The shape of the layers in the Decoder network.
+ * @param beta Weight of the latent loss in the total VAE loss. Default is 1.
```
Comment (adarob) [inline music/src/music_vae/midi_me.ts]: Weight of the *variational* loss.

File: music/src/music_vae/midi_me.ts [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+}
+
+/**
+ * An interface for providing configurable properties to the MidiMe model.
+ * @param input_size The shape of the VAE input. Since the inputs to this
+ * VAE are actually latent vectors from MusicVAE, then this number should be
+ * equal to the number of latent variables used by MusicVAE (`zDims`). The
+ * default is 256.
+ * @param output_size The size of the output. The default is 4.
+ * @param encoder_layers The shape of the layers in the Encoder network.
```
Comment (adarob) [inline music/src/music_vae/midi_me.ts]: mention defaults

File: music/src/music_vae/midi_me.ts [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+   */
+  initialize() {
+    this.dispose();
+
+    const startTime = performance.now();
+    const z = tf.input({shape: [this.config['input_size']]});
+
+    // Encoder model, goes from the original input, returns an output.
+    this.encoder = this.getEncoder(z);
+    const [sampleZ, , ] = this.encoder.apply(z) as tf.SymbolicTensor[];
```
Comment (adarob) [inline music/src/music_vae/midi_me.ts]: Maybe we should call sampleZ something like zPrime? It should be more clear that this is the latent vector for the MidiMe VAE.

File: music/src/music_vae/midi_me.ts [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+   * @param data A `Tensor` of shape `[_, this.config['output_size']]`.
+   * @param callback A function to be called at the end of every
+   * training epoch, containing the training errors for that epoch.
+   */
+  async train(xTrain: tf.Tensor, callback?: Function) {
+    const startTime = performance.now();
+    this.trained = false;
+
+    // TODO(notwaldorf): If there's a ton of data we should consider batching.
+    const optimizer = tf.train.adam();
```
Comment (adarob) [inline music/src/music_vae/midi_me.ts]: *insert adam joke*

File: music/src/music_vae/midi_me.ts [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+
+    for (let e = 0; e < this.config['epochs']; e++) {
```
Comment (adarob) [inline music/src/music_vae/midi_me.ts]: nit: can't you do this.config.epochs?

File: music/src/music_vae/midi_me.ts (hunk 12 above)
Comment (adarob) [reply music/src/music_vae/midi_me.ts]: Actually, let's just rename `z` -> `x`

File: music/src/music_vae/midi_me.ts [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+                  {units: this.config['decoder_layers'][i], activation: 'relu'})
+              .apply(x) as tf.SymbolicTensor;
+    }
+    const mu = this.getAffineLayers(x, this.config['input_size'], z, false) as
+        tf.SymbolicTensor;
+    return tf.model({inputs: z, outputs: mu, name: 'decoder'});
+  }
+
+  private loss(
+      muZ: tf.Tensor, sigmaZ: tf.Tensor, yPred: tf.Tensor, yTrue: tf.Tensor):
```
Comment (adarob) [inline music/src/music_vae/midi_me.ts]: zMu, zSigma

File: music/src/music_vae/midi_me.ts (hunk 13 above)
Comment (notwaldorf) [reply music/src/music_vae/midi_me.ts]: 🙅‍♀️

File: music/demos/midi_me.html (hunk 2 above)
Comment (notwaldorf) [reply music/demos/midi_me.html]: I added both Adam's words, and changed the demo to actually allow uploading multiple files. WDYT?

File: music/demos/common_graph.ts (hunk 3 above)
Comment (notwaldorf) [reply music/demos/common_graph.ts]: goes to show when I started this 🤦‍♀️

**Comment by notwaldorf** [comment]
I pinned the `tfjs`  version to 1.0.2 for now because updating to 1.1.0 is really bad -- the nll multiplication step gives NaNs even though it shouldn't (working on figuring that out with Nikhil)

File: music/demos/midi_me.ts [hunk 16]
```diff
@@ -16,41 +16,75 @@
...
 
 // tslint:disable-next-line:max-line-length
-import {CHECKPOINTS_DIR, visualizeNoteSeqs, writeMemory, writeTimer} from './common';
+import {CHECKPOINTS_DIR, visualizeNoteSeqs, writeTimer} from './common';
 import {updateGraph} from './common_graph';
 
 const MEL_CKPT = `${CHECKPOINTS_DIR}/music_vae/mel_2bar_small`;
-const BARS = 2;
+const TRIO_CKPT = `${CHECKPOINTS_DIR}/music_vae/trio_4bar`;
+const BARS = 4;
```
Comment (adarob) [inline music/demos/midi_me.ts]: doesn't this depend on the model?

Comment (notwaldorf) [reply music/demos/midi_me.ts]: oops, good catch. fixed.
